stage1/init/common/pod.go

```diff
@@ -938,18 +923,76 @@ func GetMachineID(p *stage1commontypes.Pod) string {
 	return "rkt-" + p.UUID.String()
 }
 
-// GetAppCapabilities is a filter which returns a string slice of valid Linux capabilities
+// getAppCapabilities is a filter which returns a string slice of valid Linux capabilities
 // It requires list of available isolators
-func GetAppCapabilities(isolators types.Isolators) []string {
-	var caps []string
+func getAppCapabilities(isolators types.Isolators) ([]string, error) {
+	var retainCaps []string
+	var removeCaps []string
 
+	// Default caps defined in
+	// https://github.com/appc/spec/blob/master/spec/ace.md#linux-isolators
+	appDefaultCapabilities := []string{
+		"CAP_AUDIT_WRITE",
+		"CAP_CHOWN",
+		"CAP_DAC_OVERRIDE",
+		"CAP_FSETID",
+		"CAP_FOWNER",
+		"CAP_KILL",
+		"CAP_MKNOD",
+		"CAP_NET_RAW",
+		"CAP_NET_BIND_SERVICE",
+		"CAP_SETUID",
+		"CAP_SETGID",
+		"CAP_SETPCAP",
+		"CAP_SETFCAP",
+		"CAP_SYS_CHROOT",
+	}
+
+	// Iterate over the isolators defined in
+	// https://github.com/appc/spec/blob/master/spec/ace.md#linux-isolators
+	// Only read the capababilities isolators:
+	// - os/linux/capabilities-retain-set
+	// - os/linux/capabilities-remove-set
 	for _, isolator := range isolators {
-		if capSet, ok := isolator.Value().(types.LinuxCapabilitiesSet); ok &&
-			isolator.Name == types.LinuxCapabilitiesRetainSetName {
-			caps = append(caps, parseLinuxCapabilitiesSet(capSet)...)
+		if capSet, ok := isolator.Value().(types.LinuxCapabilitiesSet); ok {
+			switch isolator.Name {
+			case types.LinuxCapabilitiesRetainSetName:
+				retainCaps = append(retainCaps, parseLinuxCapabilitiesSet(capSet)...)
+			case types.LinuxCapabilitiesRevokeSetName:
+				removeCaps = append(removeCaps, parseLinuxCapabilitiesSet(capSet)...)
+			}
+		}
+	}
+
+	// appc/spec does not allow to have both the retain set and the remove
+	// set defined.
+	if len(retainCaps) > 0 && len(removeCaps) > 0 {
+		return nil, errors.New("cannot have both os/linux/capabilities-retain-set and os/linux/capabilities-remove-set")
+	}
+
+	// Neither the retain set or the remove set are defined
+	if len(retainCaps) == 0 && len(removeCaps) == 0 {
+		return appDefaultCapabilities, nil
+	}
+
+	if len(retainCaps) > 0 {
+		return retainCaps, nil
+	}
+
+	if len(removeCaps) == 0 {
+		panic("logic error with RetainSet and RemoveSet")
```

Comment:
@iaguis: I meant that `logic error` isn't a useful message, but your response is much better